### PR TITLE
Fix makefile.global for apache2 module name patch

### LINKED ISSUE #1060: PHP 7.4.0 cannot be compiled with +apxs2 in Ubuntu 19.10
### Output

```
+--------------------------------------------------------------------+
|                        *** WARNING ***                             |
|                                                                    |
| You have built PHP for Apache's current non-threaded MPM.          |
| If you change Apache to use a threaded MPM you must reconfigure    |
| PHP with --enable-maintainer-zts                                   |
config.status: creating main/build-defs.h
config.status: creating scripts/phpize
config.status: creating scripts/man1/phpize.1
config.status: creating scripts/php-config
config.status: creating scripts/man1/php-config.1
config.status: creating sapi/cli/php.1
config.status: creating sapi/phpdbg/phpdbg.1
config.status: creating sapi/cgi/php-cgi.1
config.status: creating ext/phar/phar.1
config.status: creating ext/phar/phar.phar.1
config.status: creating main/php_config.h
config.status: main/php_config.h is unchanged
config.status: executing default commands

+--------------------------------------------------------------------+
| License:                                                           |
| This software is subject to the PHP License, available in this     |
| distribution in the file LICENSE. By continuing this installation  |
| process, you are bound by the terms of this license agreement.     |
| If you do not agree with the terms of this license, you must abort |
| the installation process at this point.                            |
+--------------------------------------------------------------------+

Thank you for using PHP.

make: *** No rule to make target 'libphp7.4.0.la', needed by 'all'.  Stop.
```

### Expected Result

Compiles as does PHP 7.3.12. Building without apxs2 works just fine.

### Command

```
phpbrew --debug install -j $(nproc) --mirror=https://www.php.net 7.4.0 +default +iconv +mysql +pdo +sqlite +apxs2=$(which apxs2) +opcache +exif -- --with-mysql-sock=/var/run/mysqld/mysqld.sock
```

### Build Log

    https://gist.github.com/mahagr/6ab99bea315056a70aa102c08aded21a

### Platform

Ubuntu 19.10 (with PHP 7.4 dependencies installed)

Running PHP: 

PHP 7.3.12 / 7.4.0 (phpbrew)

Installing PHP:

PHP 7.3.12 / 7.4.0 (phpbrew)

### PR DESCRIPTION
**Pull request by markwu**
Fix Apache2 module name patch for php 7.4 and older version. I just bring $build and $logger arguments to rules(), then I can get different rules set for different version

Fixes #1060.

### CODE REVIEW

File: src/PhpBrew/PatchKit/Patch.php [hunk 1]
```diff
@@ -24,15 +24,15 @@ abstract class Patch
      *
      * @return array<PatchRule>
      */
-    abstract public function rules();
+    abstract public function rules(Buildable $build, Logger $logger);
```
Comment (morozov) [inline src/PhpBrew/PatchKit/Patch.php]: Please revert and do as suggested in https://github.com/phpbrew/phpbrew/issues/1040#issuecomment-562217403.

Comment (markwu) [reply src/PhpBrew/PatchKit/Patch.php]: Done.

**Comment by morozov** [comment]
@markwu please make sure you update the tests accordingly. The patch now requires a version.

File: src/PhpBrew/Patches/Apache2ModuleNamePatch.php [hunk 2]
```diff
@@ -50,7 +57,14 @@ public function rules()
                 'libs/libphp\$PHP_VERSION.$1'
             );
 
-        $rules[] = RegExpPatchRule::files(array('Makefile.global'))
+        // For php 7.4.0 and newer, patch 'build/Makefile.global', or patch 'Makefile.global'
```
Comment (morozov) [inline src/PhpBrew/Patches/Apache2ModuleNamePatch.php]: This comment is redundant. This is exactly what the code below reads like. The comment should be about why we're doing this, possibly with external references, not what we're doing. Please remove or rework.

Comment (markwu) [reply src/PhpBrew/Patches/Apache2ModuleNamePatch.php]: Okay.

**Comment by markwu** [comment]
@morozov How do I run the test to make sure everything okay?

File: src/PhpBrew/Patches/Apache2ModuleNamePatch.php (hunk 2 above)
Comment (markwu) [reply src/PhpBrew/Patches/Apache2ModuleNamePatch.php]: Done

**Comment by morozov** [comment]
@markwu, first you need to do `composer install` if you haven't already, then you do `vendor/bin/phpunit`.

Some existing tests are ridiculously slow and should be reworked, so if the whole suite takes too long, you can run just the failing test and make sure it passes:
```
$ vendor/bin/phpunit tests/PhpBrew/Patches/Apache2ModuleNamePatchTest.php

PHPUnit 4.8.36 by Sebastian Bergmann and contributors.

Runtime:	PHP 7.3.12 with Xdebug 2.8.0
Configuration:	/home/morozov/Projects/phpbrew/phpunit.xml.dist

E

Time: 126 ms, Memory: 8.00MB

There was 1 error:

1) PhpBrew\Tests\Patches\Apache2ModuleNamePatchTest::testPatch
ArgumentCountError: Too few arguments to function PhpBrew\Patches\Apache2ModuleNamePatch::__construct(), 0 passed in /home/morozov/Projects/phpbrew/tests/PhpBrew/Patches/Apache2ModuleNamePatchTest.php on line 35 and exactly 1 expected

/home/morozov/Projects/phpbrew/src/PhpBrew/Patches/Apache2ModuleNamePatch.php:14
/home/morozov/Projects/phpbrew/tests/PhpBrew/Patches/Apache2ModuleNamePatchTest.php:35

FAILURES!
Tests: 1, Assertions: 1, Errors: 1.
```

**Comment by markwu** [comment]
Okay, I already add a test for this PR, test both `5.5.17` and `7.4.0`.

**Comment by markwu** [comment]
BTW, there are lot of error notice when generating the testing report.

```
Generating code coverage report in Clover XML format ... done

Generating Crap4J report XML file ... done

Generating code coverage report in HTML format ...PHP Warning:  count(): Parameter must be an array or an object that implements Countable in /Users/mark/Temp/phpbrew/vendor/phpunit/php-code-coverage/src/CodeCoverage/Report/HTML/Renderer/File.php on line 313
PHP Stack trace:
PHP   1. {main}() /Users/mark/Temp/phpbrew/vendor/phpunit/phpunit/phpunit:0
PHP   2. PHPUnit_TextUI_Command::main() /Users/mark/Temp/phpbrew/vendor/phpunit/phpunit/phpunit:52
PHP   3. PHPUnit_TextUI_Command->run() /Users/mark/Temp/phpbrew/vendor/phpunit/phpunit/src/TextUI/Command.php:100
PHP   4. PHPUnit_TextUI_TestRunner->doRun() /Users/mark/Temp/phpbrew/vendor/phpunit/phpunit/src/TextUI/Command.php:149
PHP   5. PHP_CodeCoverage_Report_HTML->process() /Users/mark/Temp/phpbrew/vendor/phpunit/phpunit/src/TextUI/TestRunner.php:501
PHP   6. PHP_CodeCoverage_Report_HTML_Renderer_File->render() /Users/mark/Temp/phpbrew/vendor/phpunit/php-code-coverage/src/CodeCoverage/Report/HTML.php:121
PHP   7. PHP_CodeCoverage_Report_HTML_Renderer_File->renderSource() /Users/mark/Temp/phpbrew/vendor/phpunit/php-code-coverage/src/CodeCoverage/Report/HTML/Renderer/File.php:82
PHP Warning:  count(): Parameter must be an array or an object that implements Countable in /Users/mark/Temp/phpbrew/vendor/phpunit/php-code-coverage/src/CodeCoverage/Report/HTML/Renderer/File.php on line 313

```

**Comment by morozov** [comment]
Please rebase. I cleaned up the test suite a bit yesterday. The rest is inevitable due to the outdated PHPUnit used with a new PHP.

File: tests/PhpBrew/Patches/Apache2ModuleNamePatchTest.php [hunk 3]
```diff
@@ -12,34 +12,47 @@
...
+        $patch = new Apache2ModuleNamePatch($targetVersion);
         $matched = $patch->match($build, $logger);
         $this->assertTrue($matched, 'patch matched');
         $patchedCount = $patch->apply($build, $logger);
-        $this->assertEquals(107, $patchedCount);
 
-        $sourceExpectedDirectory = getenv('PHPBREW_EXPECTED_PHP_DIR') . DIRECTORY_SEPARATOR . '5.5.17-apxs-patch';
-        $this->assertFileEquals($sourceExpectedDirectory . '/Makefile.global', $sourceDirectory . '/Makefile.global');
+        $sourceExpectedDirectory = getenv('PHPBREW_EXPECTED_PHP_DIR') . DIRECTORY_SEPARATOR . $targetVersion.'-apxs-patch';
+        if (version_compare($targetVersion, '7.4') >= 0) {
```
Comment (morozov) [inline tests/PhpBrew/Patches/Apache2ModuleNamePatchTest.php]: Please avoid conditionals in tests. Instead, you can pass the expected patched count as a parameter together with the version. Also, you can replace the newly introduced test methods with a `@dataProvider`.

Comment (markwu) [reply tests/PhpBrew/Patches/Apache2ModuleNamePatchTest.php]: Done. Use @dataProvider as you requested, instead of conditionals in tests.

**Comment by morozov** [comment]
Thank you @markwu!

**Pull request merged**